### ISSUE TITLE
[Docs error] Wrong BlocListener signature in Flutter Bloc Core Concepts page

### ISSUE THREAD

**Issue by fabryx92** [felangel/bloc#553]
This is what in the page https://github.com/felangel/bloc/blob/master/docs/flutterbloccoreconcepts.md#blocprovider:

```dart
BlocListener<BlocA, BlocAState>(
  listener: (context, state) {
    // do stuff here based on BlocA's state
  },
  child: Container(),
)
```

Shouldn't be like this?

```dart
BlocListener<BlocAEvent, BlocAState>(
  listener: (context, state) {
    // do stuff here based on BlocA's state
  },
  child: Container(),
)
```

**Comment by felangel** [comment]
Hi @fabryx92 👋 
Thanks for opening an issue!

The documentation is correct in this case. I think you are probably still using an older version of `flutter_bloc` but in v0.20.0 the signature changed in order to allow for automatic bloc lookup 👍 

Hope that helps!

**Comment by jwehrle** [comment]
I'm having a similar problem with the signature. I'm using version 7.0.0 and I want to use BlocListener for a Bloc - of the type that has events and states. But the BlocListener in my plugin will only listen to type BlocBase.

I suppose that instead of BlocListener I could use a StreamBuilder and use the stream of states from my Bloc but this seems weird to me. I know that I have used BlocListener to listen to Bloc<E, S>.

(I don't actually use BlocBase<S> because I need the events)

**Comment by felangel** [comment]
@jwehrle can you provide a small snippet that illustrates the problem you’re facing with v7.0.0? Thanks!

**Comment by jwehrle** [comment]
`@override
  Widget build(BuildContext context) {
    return BlocListener<SomethingBloc, SomethingState>(
      listener: (context, state) {},
      bloc: _getBloc(context),
      child:... `
      
The error: 'SomethingBloc' doesn't extend 'B extends BlocBase\<S>'. 

Signature of SomethingBloc
`class SomethingBloc extends Bloc<SomethingEvent, SomethingState>`

**Comment by felangel** [comment]
@jwehrle what version of bloc are you using? I’m guessing you’re not using v7.0.0 and you just need to upgrade.

**Comment by jwehrle** [comment]
Sorry about the formatting of the code block, not sure what happened.

**Comment by felangel** [comment]
@jwehrle what version of bloc and flutter_bloc are you using? I’m guessing you just need to upgrade your bloc version to v7.0.0

**Comment by jwehrle** [comment]
My pubspec.yaml has
bloc: ^7.0.0
flutter_bloc: ^7.0.0

Maybe I need to clean? A cache error?

**Comment by felangel** [comment]
> My pubspec.yaml has
> bloc: ^7.0.0
> flutter_bloc: ^7.0.0
> 
> Maybe I need to clean? A cache error?

Yeah I would clear your pub cache and make sure you’re using v7.0.0 of each library. Also an IDE restart sometimes helps 😅

**Comment by jwehrle** [comment]
Frustrating to get Android Studio to refresh cache. Pub cache does not want to clear with the clean command or a restart.

**Comment by felangel** [comment]
> Frustrating to get Android Studio to refresh cache. Pub cache does not want to clear with the clean command or a restart.

Try a `pub cache repair`

**Comment by jwehrle** [comment]
Alright, I've run pub cache repair (had to also update dart, did that, all good) and restarted (with invalidated caches) and still getting the error. Checked my plugins and both bloc and flutter_bloc are listed as version 7.0.0

Not sure what's left to do.

**Comment by jwehrle** [comment]
Under Dart Packages:
"bloc-7.0.0" and further down, "flutter_bloc-7.0.0"

**Comment by felangel** [comment]
@jwehrle can you create a reproduction sample?

**Comment by jwehrle** [comment]
Sample (works as expected):

main
```import 'package:flutter/material.dart';
import 'package:bloctest/test_bloc.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter FUN',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: BlocProvider<FunBloc>(
        create: (BuildContext context) {
          return FunBloc(Standby());
        },
        child: MyHomePage(title: 'Flutter FUN Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  void _showFun(ShowingFun state) async {
    await showDialog(
        context: context,
        builder: (context) {
          return AlertDialog(
            title: Text('Fun!!'),
            content: Text(state.funString),
            actions: [
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: Text('Fun Done'),
              ),
            ],
          );
        });
    BlocProvider.of<FunBloc>(context).add(EndFun());
  }

  @override
  Widget build(BuildContext context) {
    return BlocListener<FunBloc, TestState>(
      bloc: BlocProvider.of<FunBloc>(context),
      listener: (context, state) {
        if (state is ShowingFun) {
          _showFun(state);
        }
      },
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: Text(
            'Press button for FUN!',
            style: Theme.of(context).textTheme.headline4,
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            BlocProvider.of<FunBloc>(context).add(ShowFun());
          },
          tooltip: 'Increment',
          child: Icon(Icons.add),
        ),
      ),
    );
  }
}
```
fun_bloc
```import 'dart:math';

import 'package:bloc/bloc.dart';
import 'package:equatable/equatable.dart';

class FunBloc extends Bloc<FunEvent, FunState> {
  FunBloc(Standby initialState) : super(initialState);

  @override
  Stream<FunState> mapEventToState(FunEvent event) async* {
    if (event is ShowFun) {
      yield ShowingFun(getNextFun());
    }
    if (event is EndFun) {
      yield Standby();
    }
  }
}

String getNextFun() {
  return fun[Random().nextInt(5)];
}

// from https://randomwordgenerator.com/
List<String> fun = List.unmodifiable([
  'Insert',
  'Invisible',
  'Rally',
  'Bear',
  'Underline',
]);

abstract class FunEvent extends Equatable {}

class ShowFun extends FunEvent {
  @override
  List<Object?> get props => [];
}

class EndFun extends FunEvent {
  @override
  List<Object?> get props => [];
}

abstract class FunState extends Equatable {}

class Standby extends FunState {
  @override
  List<Object?> get props => [];
}

class ShowingFun extends FunState {
  final String funString;

  ShowingFun(this.funString);

  @override
  List<Object?> get props => [funString];
}
```
pubspec.yaml
```name: bloctest
description: A new Flutter application.

# The following line prevents the package from being accidentally published to
# pub.dev using `pub publish`. This is preferred for private packages.
publish_to: 'none' # Remove this line if you wish to publish to pub.dev

# The following defines the version and build number for your application.
# A version number is three numbers separated by dots, like 1.2.43
# followed by an optional build number separated by a +.
# Both the version and the builder number may be overridden in flutter
# build by specifying --build-name and --build-number, respectively.
# In Android, build-name is used as versionName while build-number used as versionCode.
# Read more about Android versioning at https://developer.android.com/studio/publish/versioning
# In iOS, build-name is used as CFBundleShortVersionString while build-number used as CFBundleVersion.
# Read more about iOS versioning at
# https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html
version: 1.0.0+1

environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter

  bloc: ^7.0.0
  flutter_bloc: ^7.0.0
  equatable: ^2.0.0

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^1.0.2

dev_dependencies:
  flutter_test:
    sdk: flutter

# For information on the generic Dart part of this file, see the
# following page: https://dart.dev/tools/pub/pubspec

# The following section is specific to Flutter.
flutter:

  # The following line ensures that the Material Icons font is
  # included with your application, so that you can use the icons in
  # the material Icons class.
  uses-material-design: true

  # To add assets to your application, add an assets section, like this:
  # assets:
  #   - images/a_dot_burr.jpeg
  #   - images/a_dot_ham.jpeg

  # An image asset can refer to one or more resolution-specific "variants", see
  # https://flutter.dev/assets-and-images/#resolution-aware.

  # For details regarding adding assets from package dependencies, see
  # https://flutter.dev/assets-and-images/#from-packages

  # To add custom fonts to your application, add a fonts section here,
  # in this "flutter" section. Each entry in this list should have a
  # "family" key with the font family name, and a "fonts" key with a
  # list giving the asset and other descriptors for the font. For
  # example:
  # fonts:
  #   - family: Schyler
  #     fonts:
  #       - asset: fonts/Schyler-Regular.ttf
  #       - asset: fonts/Schyler-Italic.ttf
  #         style: italic
  #   - family: Trajan Pro
  #     fonts:
  #       - asset: fonts/TrajanPro.ttf
  #       - asset: fonts/TrajanPro_Bold.ttf
  #         weight: 700
  #
  # For details regarding fonts from package dependencies,
  # see https://flutter.dev/custom-fonts/#from-packages
```

**Comment by jwehrle** [comment]
So obviously there's still something weird with my project. :(

**Comment by jwehrle** [comment]
OK, I tried something weird that should not have worked but seems to have. I copied the bloc_listener file into project and simply replaced references to BlocBase with Bloc. Then imported (with namespace myBloc) to file I wanted to use it in and Android Studio was happy. THEN, I removed the namespace modifier myBloc and Android Studio is still happy.

Which I think tells me that this was maybe all a problem in the IDE's imagination? That maybe it couldn't tell that Bloc extends BlocBase? I'm not sure. Thanks @felangel  for discussing this issue with me.
